rfd/0073-discover.md

```diff
@@ -0,0 +1,337 @@
+
+- Works with both self-hosted edition and Teleport Cloud.
+- Focuses on discovering resources in AWS, other clouds to follow.
+- Covers currently supported cloud resources: EC2 instances and RDS databases.
+- The phase 1 of the implementation will only support EC2 instances and SSH.
+
+## Prerequisites
+
+In order to use `teleport discover` the user will need:
+
+- Running Auth and Proxy, or a Cloud account.
+- An AWS EC2 instance to run `teleport discover` command from. The node must have
+  IAM identity allowing it to perform discovery operations in AWS account.
+
+## Command
+
+The goal for `teleport discover` is to eventually support multiple cloud
+providers. Given that different providers often have different concepts and
+different names for similar things, it may be hard to unify them all under a
+single `teleport discover` command without making UX confusing, having
+overlapping or prefixed flags, etc.
+
+Hence, the proposal is to have a family of subcommands for different cloud
+providers:
+
+```
+$ teleport discover aws ...
+$ teleport discover gcp ...
+```
+
+This RFD will focus on `teleport discover aws`.
+
+## Usage
+
+The default mode of operation of the `teleport discover` command is to support
+the UI wizard users will use to join their resources to the cluster.
+
+Users will go through the following flow in the wizard-style dialog in the
+Teleport Web UI:
+
+- Select which environment to look for resources in. E.g. self-hosted (which
+  will use existing automatic join scipts), AWS or other cloud (which will use
+  the discover flow), as well as regions.
+- Select types of resources to import. For AWS users will be offered all types
+  we currently support auto-discovery for: EC2, RDS/Aurora, Redshift, Elasticache,
+  MemoryDB. The dialog may optionally allow to specify tags to filter the
+  resources by.
+- Based on the selected resources, display appropriate `teleport discover aws`
+  command to run and the IAM policy required for the command to successfully
+  discover and join resources.
+
+Users will launch the `teleport discover aws` command on an AWS instance that
+has IAM permissions required to discover and connect resoures. We'll detail
+those permissions later.
+
+The command will:
+
+1. Join the Teleport cluster as a "discover" agent using the same cluster join
+   mechanism as any other agent. Join token will be generated by the server
+   and included in the CLI command displayed to a user.
+2. Using the node's IAM credentials, search the cloud account for the requested
+   resources according to the command's filters.
+3. Once discovered, the agent will start heartbeating the discovered resources
+   to the cluster with TTL using the same heartbeat mechanism as any other
+   type of resource. Heartbeating resources to the cluster will allow the UI
+   to pick them up and show to the user.
+4. Wait for the user to inspect discovered resources in the UI wizard and select
+   those they would like to connect.
```

Comment:
Similar to above, we should preselect all discovered resources and have users opt out instead of opt in.